RIOT-OS/RIOT
# wolfSSL pkg addition with examples

### Contribution description

This patch superseeds PR #9894  and contains wolfSSL support in a RIOT pkg.

This patch includes:
  - independent modules that can be selected and integrated separately into applications
  - single cryptography algorithm selectable as separate modules
  - Support for TLS over TCP/IP using **sockets**
  - Support for DTLS over UDP/IP using **sock_udp** interface

### Modules
**wolfCrypt**:
 - wolfcrypt_poly1305
 - wolfcrypt_chacha
 - wolfcrypt_ed25519
 - wolfcrypt_aes
 - wolfcrypt_pwdbased
 - wolfcrypt_asn
 - wolfcrypt_random
 - wolfcrypt_* (e.g. ecc, rsa, sha3, ...)

**wolfSSL**:
 - wolfssl_tls
 - wolfsll_tls13
 - wolfssl_dtls
 - wolfssl_* (e.g. ocsp, crl, ...)


### Testing procedure: sock_udp

The example `dtls-wolfssl` provided uses similar shell commands to those in dtls-echo:

`dtlss`
 start dtls server on port 11111

`dtlsc addr`
start dtls client demo connection to server at address *addr*

`wolftest`
 run all sanity checks to ensure that the crypto engine is working on the target

#### Compile and run dtls-wolfssl

wolfSSL works on any 32-bit target with support for `gnrc_networking`. 

The test procedure for native riot-to-riot test is the following:

#### Prepare the bridge interface linking two tuntap:

    ./../../dist/tools/tapsetup/tapsetup --create 2

#### Run the server
```
    $ make all; PORT=tap1 make term
    > ifconfig
```
*copy the server address*

```
   > dtlss
```
#### Run the client
```
    $ PORT=tap0 make term
    > dtlsc <IPv6's server address[%netif]>
```

### Testing procedure: full TCP/IP TLS demo

The examples `wolfssl-client` and `wolfssl-server` provide a TLS demo with full TCP/IP connectivity, and supports all 32-bit platforms, used in combination with the lwIP package. The TCP socket support depends on `lwip_sock_tcp`.

### Issues/PRs references

This PR includes all previous work made in #9894 and #7348, rebased as single commit on the current master.

The wolfSSL library package temporarly lives in the feature branch https://github.com/danielinux/wolfssl/tree/riot-os until the new callbacks for GNRC are merged in.

### Acknowledgements

Thanks to @cladmi for reviewing the modules integration and helping with modularization of the single components.

Thanks to @bergzand for testing wolfcrypt_ed25519 and suggesting code-size optimization settings.

Please consider for inclusion.

Any feedback is welcome.

-- @danielinux -- @kaleb-himes -- @wolfssl
